### PR TITLE
Add some type hints to datastore

### PR DESCRIPTION
**Pull request by dklimpel**
### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

Signed-off-by: Dirk Klimpel dirk@klimpel.org

### CODE REVIEW

File: synapse/storage/databases/main/group_server.py [hunk 1]
```diff
@@ -528,15 +546,15 @@ def _get_all_groups_for_user_txn(txn):
             "get_all_groups_for_user", _get_all_groups_for_user_txn
         )
 
-    async def get_groups_changes_for_user(self, user_id, from_token, to_token):
+    async def get_groups_changes_for_user(self, user_id: str, from_token, to_token):
```
Comment (clokep) [inline synapse/storage/databases/main/group_server.py:549]: I think this is:

```suggestion
    async def get_groups_changes_for_user(self, user_id: str, from_token: int, to_token: int) -> List[JsonDict]:
```

(And the same return type on the inner function.)

File: synapse/storage/databases/main/group_server.py [hunk 2]
```diff
@@ -583,12 +601,12 @@ async def get_all_groups_changes(
...
 
         last_id = int(last_id)
-        has_changed = self._group_updates_stream_cache.has_any_entity_changed(last_id)
+        has_changed = self._group_updates_stream_cache.has_any_entity_changed(last_id)  # type: ignore[attr-defined]
 
         if not has_changed:
             return [], current_id, False
 
-        def _get_all_groups_changes_txn(txn):
+        def _get_all_groups_changes_txn(txn: LoggingTransaction):
```
Comment (clokep) [inline synapse/storage/databases/main/group_server.py:609]:
```suggestion
        def _get_all_groups_changes_txn(txn: LoggingTransaction) -> Tuple[List[Tuple[int, tuple]], int, bool]:
```

File: synapse/storage/databases/main/group_server.py [hunk 3]
```diff
@@ -766,7 +784,7 @@ def _add_room_to_summary_txn(
                     "category_id": category_id,
                     "room_id": room_id,
                 },
-                values=to_update,
+                updatevalues=to_update,
```
Comment (clokep) [inline synapse/storage/databases/main/group_server.py:787]: Oooooo, a real bug fix! 🎉 Nice find!

File: synapse/storage/databases/main/monthly_active_users.py [hunk 4]
```diff
@@ -141,12 +142,12 @@ async def user_last_seen_monthly_active(self, user_id: str) -> Optional[int]:
...
 
     @wrap_as_background_process("reap_monthly_active_users")
-    async def reap_monthly_active_users(self):
+    async def reap_monthly_active_users(self) -> None:
         """Cleans out monthly active user table to ensure that no stale
         entries exist.
         """
 
-        def _reap_users(txn, reserved_users):
+        def _reap_users(txn: LoggingTransaction, reserved_users: List[str]):
```
Comment (clokep) [inline synapse/storage/databases/main/monthly_active_users.py:150]:
```suggestion
        def _reap_users(txn: LoggingTransaction, reserved_users: List[str]) -> None:
```

**Review by clokep — DISMISSED** [review]
Looks good! I found a few spots that I think we could add return types to.

File: synapse/storage/databases/main/group_server.py [hunk 5]
```diff
@@ -750,7 +768,7 @@ def _add_room_to_summary_txn(
                 WHERE group_id = ? AND category_id = ?
             """
             txn.execute(sql, (group_id, category_id))
-            (order,) = txn.fetchone()
+            (order,) = cast(Tuple[Optional[int]], txn.fetchone())
```
Comment (DMRobertson) [inline synapse/storage/databases/main/group_server.py:775]: Is this correct?

- `MAX(X)` might return null if there are no `room_orders` selected,
- but if so, `COALESCE(NULL, 0)` will give us a nonnull value (`0`).

So think `order: int` rather than `order: Optional[int]`.

File: synapse/storage/databases/main/group_server.py (hunk 3 above)
Comment (DMRobertson) [inline synapse/storage/databases/main/group_server.py:791]: This is alarming---is this a bugfix? If so, any idea why we haven't noticed sooner?

File: synapse/storage/databases/main/group_server.py [hunk 6]
```diff
@@ -989,7 +1007,7 @@ def _add_user_to_summary_txn(
                 WHERE group_id = ? AND role_id = ?
             """
             txn.execute(sql, (group_id, role_id))
-            (order,) = txn.fetchone()
+            (order,) = cast(Tuple[Optional[int]], txn.fetchone())
```
Comment (DMRobertson) [inline synapse/storage/databases/main/group_server.py:1014]: Same as a comment above: I think order is never NULL?

File: synapse/storage/databases/main/monthly_active_users.py [hunk 7]
```diff
@@ -210,10 +211,10 @@ def _reap_users(txn, reserved_users):
             # is racy.
             # Have resolved to invalidate the whole cache for now and do
             # something about it if and when the perf becomes significant
-            self._invalidate_all_cache_and_stream(
+            self._invalidate_all_cache_and_stream(  # type: ignore[attr-defined]
```
Comment (DMRobertson) [inline synapse/storage/databases/main/monthly_active_users.py:214]: Urgh. Fixing this properly would require us to inherit from `CacheInvalidationWorkerStore`. [I've been down that road](https://github.com/matrix-org/synapse/pull/11354), and it's not pretty.

File: synapse/storage/databases/main/monthly_active_users.py [hunk 8]
```diff
@@ -261,10 +264,10 @@ def _initialise_reserved_users(self, txn, threepids):
         #    run this at startup?
 
         for tp in threepids:
-            user_id = self.get_user_id_by_threepid_txn(txn, tp["medium"], tp["address"])
+            user_id = self.get_user_id_by_threepid_txn(txn, tp["medium"], tp["address"])  # type: ignore[attr-defined]
 
             if user_id:
-                is_support = self.is_support_user_txn(txn, user_id)
+                is_support = self.is_support_user_txn(txn, user_id)  # type: ignore[attr-defined]
```
Comment (DMRobertson) [inline synapse/storage/databases/main/monthly_active_users.py:270]: Could try making this class inherit from `RegistrationWorkerStore` to make this and the cache invalidation methods visible. It might give you MRO errors though---fine to leave as it is.

File: mypy.ini [hunk 9]
```diff
@@ -87,9 +84,6 @@ exclude = (?x)
    |tests/state/test_v2.py
    |tests/storage/test_background_update.py
    |tests/storage/test_base.py
-   |tests/storage/test_client_ips.py
-   |tests/storage/test_database.py
-   |tests/storage/test_event_federation.py
```
Comment (DMRobertson) [inline mypy.ini:92]: Is this a related change? Perhaps the datastore annotations are needed to get these tests to typecheck?

**Review by DMRobertson — COMMENTED** [review]
Whoops. Looks like Patrick and I hit a race condition.

Many thanks for this. A few comments---not all needing a change!

File: synapse/storage/databases/main/group_server.py (hunk 3 above)
Comment (dklimpel) [reply synapse/storage/databases/main/group_server.py:791]: My thought was whether it is dead code that is never called?

File: synapse/storage/databases/main/group_server.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dklimpel) [reply synapse/storage/databases/main/group_server.py]: Done.

File: synapse/storage/databases/main/group_server.py (hunk 5 above)
Comment (dklimpel) [reply synapse/storage/databases/main/group_server.py]: I have done this.

File: synapse/storage/databases/main/group_server.py (hunk 6 above)
Comment (dklimpel) [reply synapse/storage/databases/main/group_server.py]: I have done this.

File: mypy.ini (hunk 9 above)
Comment (dklimpel) [reply mypy.ini:92]: This is related to reduce the number of exceptions for mypy.
These test are ok at the moment and does not create an error in the current status.  But they are certainly not complete with annotations. Like most tests.

File: synapse/storage/databases/main/monthly_active_users.py (hunk 8 above)
Comment (DMRobertson) [reply synapse/storage/databases/main/monthly_active_users.py]: Wow, it worked!!

File: synapse/storage/databases/main/monthly_active_users.py [hunk 10]
```diff
@@ -328,15 +334,15 @@ def upsert_monthly_active_user_txn(self, txn, user_id):
             values={"timestamp": int(self._clock.time_msec())},
         )
 
-        self._invalidate_cache_and_stream(txn, self.get_monthly_active_count, ())
-        self._invalidate_cache_and_stream(
+        self._invalidate_cache_and_stream(txn, self.get_monthly_active_count, ())  # type: ignore[attr-defined]
```
Comment (DMRobertson) [inline synapse/storage/databases/main/monthly_active_users.py:337]: I think some more of these ignores are redundant after 3a7ef03b1f9b894a29820ba65e0701c4f1b97cb7. Try running `mypy --warn-unused-ignores synapse/storage/databases/main`!

File: synapse/storage/databases/main/group_server.py [hunk 11]
```diff
@@ -528,15 +546,17 @@ def _get_all_groups_for_user_txn(txn):
             "get_all_groups_for_user", _get_all_groups_for_user_txn
         )
 
-    async def get_groups_changes_for_user(self, user_id, from_token, to_token):
+    async def get_groups_changes_for_user(
+        self, user_id: str, from_token: int, to_token: int
+    ) -> List[JsonDict]:
         from_token = int(from_token)
```
Comment (DMRobertson) [inline synapse/storage/databases/main/group_server.py:552]: I think this line is now redundant?

File: synapse/storage/databases/main/group_server.py (hunk 3 above)
Comment (DMRobertson) [reply synapse/storage/databases/main/group_server.py:794]: I guess so. Still, good spot!

**Review by DMRobertson — APPROVED** [review]
Thanks Dirk---lgtm!

@clokep, since you've already taken a look at this, do you want to have another skim over the changes too?

**Comment by clokep** [comment]
> Thanks Dirk---lgtm!
> 
> @clokep, since you've already taken a look at this, do you want to have another skim over the changes too?

I'm happy if you're happy! 👍
